File: node/src/routes/prankster.js
import { Router } from 'express';

import { handleError, hasKeys, toNumber } from '../common/util.js';

const send = async (res, { exceptionMessage, statusCode, inputDelay }) => {
  const doit = () => {
    if (exceptionMessage !== undefined) {
      throw new Error(exceptionMessage);
    }
    res.status(statusCode).end();
  };

  const delay = inputDelay !== undefined
    ? toNumber(inputDelay, 'delay')
    : 0;

  if (delay === 0) {
    return doit();
  }

  return new Promise((resolve, reject) => {
    setTimeout(
      () => {
        try {
          doit();
          resolve();
        } catch (error) {
          reject(error);
        }
      },
      delay,
    );
  });
};

export default () => {
  const router = Router();

  router.post('/', handleError(async (req, res) => {
    const input = req.body;
    if (hasKeys(input)) {
      const exceptionMessage = input.exceptionMessage;
      if (exceptionMessage !== undefined) {
        if (typeof exceptionMessage !== 'string' || exceptionMessage.length === 0) {
          throw new Error(`The value of 'exceptionMessage' must be a not-empty string and is '${exceptionMessage}'.`);
        }
        return send(res, { exceptionMessage, inputDelay: input.delay });
      }

      if (input.statusCode !== undefined) {
        const statusCode = toNumber(input.statusCode, 'statusCode');
        if (statusCode < 100) {
          throw new Error(`The value of 'statusCode' must be a number greater or equal to 100 and is '${statusCode}'.`);
        }
        return send(res, { statusCode, inputDelay: input.delay });
      }
    }

    throw new Error('Either \'statusCode\' or \'exceptionMessage\' must be informed.');
  }));

  return router;
};

File: go/src/routes/prankster.go
package routes

import (
  "encoding/json"
  "fmt"
  "io/ioutil"
  "net/http"
  "time"

  "lappdog/models"
)

func PrankIt(w http.ResponseWriter, r *http.Request) {
  reqBody, err := ioutil.ReadAll(r.Body)
  if err != nil {
    sendError(w, err)
    return
  }

  var input models.PranksterInput
  err = json.Unmarshal(reqBody, &input)
  if err != nil {
    sendError(w, err)
    return
  }

  if input.ExceptionMessage != "" {
    if input.Delay > 0 {
      time.Sleep(time.Duration(input.Delay) * time.Millisecond)
    }

    sendError(w, fmt.Errorf(input.ExceptionMessage))
    return
  }

  if input.StatusCode != nil {
    code := *input.StatusCode
    if code < 100 {
      sendError(w, fmt.Errorf("The value of 'statusCode' must be a number greater or equal to 100 and is '%d'.", code))
      return
    }

    if input.Delay > 0 {
      time.Sleep(time.Duration(input.Delay) * time.Millisecond)
    }

    w.WriteHeader(code)
    return
  }

  sendError(w, fmt.Errorf("Either 'statusCode' or 'exceptionMessage' must be informed."))
}
